test/common/TestCommon.hs

```diff
@@ -149,3 +153,50 @@ qcTestRanges txt f fArgDom dsRange perturbationRange dtRange =
   forAll (choose perturbationRange) $ \perturbationRaw ->
   forAll (choose dtRange) $ \dt ->
   ioProperty $ qcPropFArg f fArgDom xyz dsRaw perturbationRaw dt
+
+----------------------------------------------------------------------------
+-- Things that have shape
+----------------------------------------------------------------------------
+
+-- | Shape of the thing, typically a list of the sizes of its dimensions.
+type ShapeL = [Int]
+
+class HasShape a where
+  shapeL :: a -> ShapeL
+
+instance (VS.Storable a) => HasShape (VS.Vector a) where
+  shapeL vec = [VS.length vec]
+
+instance HasShape (OT.Array a) where
+  shapeL = OT.shapeL
+
+instance (OS.Shape sh) => HasShape (OS.Array sh a) where
+  shapeL = OS.shapeL
+
+instance HasShape (HM.Matrix a) where
+  shapeL matrix = [HM.rows matrix, HM.cols matrix]
+
+instance {-# OVERLAPPABLE #-} (Foldable t) => HasShape (t a) where
+  shapeL = (.) (flip (:) []) length
```

Comment:
This is an `OVERLAPPABLE` instance to be used with `Foldable` as a fallback... in order to avoid typing instances for things like `(Data.Vector.Vector (Data.Vector.Storable.Vector Float))`. The one for the matrix is right above it and is stricly more specific so it will get selected, correct?

Reply:
Correct, I guess. Yes, this concrete instance is useful and the shape would be legit. I suppose that's worth the risk for when we start using any of the other dozen of matrix types.

Reply:
When we introduce new matrix types (higher dimensions?...), we can (and should) add more instances. But even if we forget to do so, the worst thing that can happen is (I think) that we mistakenly judge some `2x3` and `1x6` matrices, miraculously containing the same data, to be equal. This is highly unlikely.

Do you have any specific matrix types in mind? Perhaps we could implement those instances now.

Reply:
Massiv and/or Accelerate matrices, as well as anything that may be needed for GPU. I think it's too early to touch the former two. I agree the risk is low. What bothered me was I didn't understand the need for that instance, but now I do. :)